File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic/BoundXmlCData.cs
using System.Diagnostics;
using Microsoft.CodeAnalysis.VisualBasic.Symbols;

namespace Microsoft.CodeAnalysis.VisualBasic
{
	internal sealed class BoundXmlCData : BoundExpression
	{
		private readonly BoundLiteral _Value;

		private readonly BoundExpression _ObjectCreation;

		public BoundLiteral Value => _Value;

		public BoundExpression ObjectCreation => _ObjectCreation;

		public BoundXmlCData(SyntaxNode syntax, BoundLiteral value, BoundExpression objectCreation, TypeSymbol type, bool hasErrors = false)
			: base(BoundKind.XmlCData, syntax, type, hasErrors || BoundNodeExtensions.NonNullAndHasErrors(value) || BoundNodeExtensions.NonNullAndHasErrors(objectCreation))
		{
			_Value = value;
			_ObjectCreation = objectCreation;
		}

		[DebuggerStepThrough]
		public override BoundNode Accept(BoundTreeVisitor visitor)
		{
			return visitor.VisitXmlCData(this);
		}

		public BoundXmlCData Update(BoundLiteral value, BoundExpression objectCreation, TypeSymbol type)
		{
			if (value != Value || objectCreation != ObjectCreation || (object)type != base.Type)
			{
				BoundXmlCData boundXmlCData = new BoundXmlCData(base.Syntax, value, objectCreation, type, base.HasErrors);
				boundXmlCData.CopyAttributes(this);
				return boundXmlCData;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/BoundXmlCData.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
Imports System
Imports System.Diagnostics

Namespace Microsoft.CodeAnalysis.VisualBasic
	Friend NotInheritable Class BoundXmlCData
		Inherits BoundExpression
		Private ReadOnly _Value As BoundLiteral

		Private ReadOnly _ObjectCreation As BoundExpression

		Public ReadOnly Property ObjectCreation As BoundExpression
			Get
				Return Me._ObjectCreation
			End Get
		End Property

		Public ReadOnly Property Value As BoundLiteral
			Get
				Return Me._Value
			End Get
		End Property

		Public Sub New(ByVal syntax As SyntaxNode, ByVal value As BoundLiteral, ByVal objectCreation As BoundExpression, ByVal type As TypeSymbol, Optional ByVal hasErrors As Boolean = False)
			MyBase.New(BoundKind.XmlCData, syntax, type, If(hasErrors OrElse value.NonNullAndHasErrors(), True, objectCreation.NonNullAndHasErrors()))
			Me._Value = value
			Me._ObjectCreation = objectCreation
		End Sub

		<DebuggerStepThrough>
		Public Overrides Function Accept(ByVal visitor As BoundTreeVisitor) As BoundNode
			Return visitor.VisitXmlCData(Me)
		End Function

		Public Function Update(ByVal value As BoundLiteral, ByVal objectCreation As BoundExpression, ByVal type As TypeSymbol) As BoundXmlCData
			Dim boundXmlCDatum As BoundXmlCData
			If (value <> Me.Value OrElse objectCreation <> Me.ObjectCreation OrElse CObj(type) <> CObj(MyBase.Type)) Then
				Dim boundXmlCDatum1 As BoundXmlCData = New BoundXmlCData(MyBase.Syntax, value, objectCreation, type, MyBase.HasErrors)
				boundXmlCDatum1.CopyAttributes(Me)
				boundXmlCDatum = boundXmlCDatum1
			Else
				boundXmlCDatum = Me
			End If
			Return boundXmlCDatum
		End Function
	End Class
End Namespace